src/io/XTcpSocket.cpp

```diff
@@ -0,0 +1,193 @@
+    else
+        snprintf(lastError, 255, "%s error : %s ", func, msgbuf);
+
+#else
+    snprintf(lastError, 255, "%s error : %s ", func, strerror(errno));
+#endif
+}
+
+const char* XTcpSocket::getLastErrorStr(void) {
+    return lastError;
+}
+
+void XTcpSocket::close() {
+    // shutdown (sockfd,SHUT_RDWR );
+#ifdef _MSC_VER
+    closesocket(sockfd);
+#else
+    ::close(sockfd);
+#endif
+    sockfd = -1;
+}
+
+void XTcpSocket::abort() {
+    close();
+}
+
+int XTcpSocket::write(const QByteArray& buff, const size_t size) {
+    int ret = send(sockfd, buff.data(), size, MSG_NOSIGNAL);
+    if (ret != (int)size) {
+        FormatLastErrorStr("write");
+        return -1;
+    }
+    return ret;
+}
+
+int XTcpSocket::read(QByteArray& buff, const size_t size) {
+    int ret = recv(sockfd, buff.data(), size, MSG_WAITALL);
+    if (ret != (int)size) {
+        FormatLastErrorStr("read");
+        return -1;
+    }
+    return ret;
+}
+
+int XTcpSocket::connectToHost(const QString& host, int port) {
+    struct sockaddr_in serv;
+
+    if ((sockfd = socket(PF_INET, SOCK_STREAM, 0)) < 0) {
+        FormatLastErrorStr("socket");
+        return 0;
+    }
+
+    memset(&serv, 0, sizeof(serv));
+    serv.sin_family = AF_INET;
+    serv.sin_addr.s_addr = inet_addr(host.toStdString().c_str());
+    serv.sin_port = htons(port);
+
+    if (connect(sockfd, (struct sockaddr*)&serv, sizeof(serv)) < 0) {
+        FormatLastErrorStr("connect");
+        close();
+        return 0;
+    }
+
+    return 1;
+}
+
+int XTcpSocket::serverStart(int port) {
+    struct sockaddr_in serv;
+
+    if ((listenfd = socket(PF_INET, SOCK_STREAM, 0)) < 0) {
+        FormatLastErrorStr("socket");
+        return -1;
+    }
+
+    int reuse = 1;
+    if (setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR, (const char*)&reuse, sizeof(reuse)) < 0) {
+        FormatLastErrorStr("setsockopt(SO_REUSEADDR)");
+    }
+
+    memset(&serv, 0, sizeof(serv));
+    serv.sin_family = AF_INET;
+    serv.sin_addr.s_addr = htonl(INADDR_ANY);
+    serv.sin_port = htons(port);
+
+    if (bind(listenfd, (struct sockaddr*)&serv, sizeof(serv)) < 0) {
+        FormatLastErrorStr("bind");
+        serverClose();
+        return -1;
+    }
+
+    if (listen(listenfd, SOMAXCONN) < 0) {
+        FormatLastErrorStr("listen");
+        serverClose();
+        return -1;
+    }
+
+    return listenfd;
+}
+
+int XTcpSocket::serverAccept(void) {
+    struct sockaddr_in cli;
+#ifdef _MSC_VER
+    int clilen = sizeof(cli);
+#else
+    unsigned int clilen = sizeof(cli);
+#endif
+    sockfd = accept(listenfd, (struct sockaddr*)&cli, &clilen);
+    if (sockfd < 0) {
+        FormatLastErrorStr("accept");
+        sockfd = -1;
+    }
+    return sockfd;
+}
+
+void XTcpSocket::serverClose(void) {
+#ifdef _MSC_VER
+    closesocket(listenfd);
+#else
+    ::close(listenfd);
+#endif
+    listenfd = -1;
+}
+
+int XTcpSocket::instances = 0;
```

Comment:
For making XTcpSocket reentrant, i think it would  be good to make this an `std::atomic<int>`.

Reply:
Done. You are right.